wasm/src/main/java/com/dylibso/chicory/wasm/io/BufferWasmInputStream.java

```diff
@@ -0,0 +1,88 @@
+package com.dylibso.chicory.wasm.io;
+
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/**
+ * A WASM input stream backed by a byte buffer.
+ */
+class BufferWasmInputStream extends WasmInputStream {
+    private final ByteBuffer buffer;
```

Comment:
Have you evaluated the option of wrapping something more high level like a `ByteArrayInputStream` ?
It will make the implementation more straightforward as we won't need to deal with indexes and more

Reply:
This implementation is for when you already have a `ByteBuffer`. Using `ByteBuffer` here does not dictate anything outside of this class (note that the class is non-public). The use case is when you want to (for example) memory-map a very WASM large file, or when you have a WASM file within a buffer already (this will happen when we start talking about output).

Reply:
Note that you can also wrap an `InputStream` using a separate API and implementation.

Reply:
I would love not to have to wrangle with low-level details of the `InputStream` and underlying `ByteBuffer`, do you have an estimate on how much less performant it will be to rely on a `ByteArrayInputStream`?

For what kind of load is this going to become an issue?

Reply:
> it's not really low level in any way either
> It's quite straightforward and the implementation is pretty minimal.

I completely understand your reasons for saying so, but, I do believe that this is not completely objective, let me elaborate.

I would stay away from the complexities of having to deal directly with the `ByteBuffer` API when we can avoid so, by accepting the PR in its current form we take on the challenges of fixing bugs in this implementation.

If there are no big performance or other implications I would love to see an implementation using a higher-level Java API/construct and doing the minimal job(none to little) required to conform to the interface.

Reply:
There is no other alternative. If a WASM module is in a `ByteBuffer` (for example, due to coming from a very large memory-mapped file), then unless we copy the entire thing to the heap (potentially gigabytes?), we must read the data where it is. It's not directly about performance because this is the most direct implementation for this use case. And, there is no higher-level construct to model this concept (until Java FFM in Java 21/22/23+ which introduces `MemorySegment` - which would probably be considered even more low-level, though it is also a better fit for this task).

I understand wanting to keep the code simple. But, "keeping it simple" to me means "avoid using subtle/tricky code practices without documentation", "factor out complexity", "hide necessary complexity behind usable APIs", and things like that; it does not mean "avoid core JDK APIs that I don't understand well". As Java software engineers, understanding the core JDK APIs is an incumbent and prerequisite part of our job responsibilities. The behavioral contract of `ByteBuffer`'s methods are very clear and well-documented, and the way that this class utilizes `ByteBuffer` should also be quite clear here. There is little subtlety in this implementation; it presents a stream view over a buffer slice in the most direct way possible.

Reply:
I have no interest in continuing the discussion after this passive-aggressive comment.